File: src/lab4/UserFactory.java
package lab4;

import lab3.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class UserFactory {

    private final Map<String, UserType<?>> registry;

    public UserFactory() {
        this.registry = new HashMap<>();
        register(new IntegerType());
        register(new StringType());
        register(new DoubleType());
    }

    private void register(UserType<?> type) {
        registry.put(type.typeName(), type);
    }

    public ArrayList<String> getTypeNameList() {
        return new ArrayList<>(registry.keySet());
    }

    public UserType<?> getBuilderByName(String name) {
        UserType<?> type = registry.get(name);
        if (type == null) {
            throw new IllegalArgumentException("Unknown type: " + name);
        }
        return type;
    }
}

File: src/lab5scala/UserFactory.scala
package lab5scala

import scala.collection.mutable

class UserFactory {
  
  private val registry: mutable.HashMap[String, UserType[?]] = mutable.HashMap.empty
  
  register(new IntegerType())
  register(new StringType())
  register(new DoubleType())

  private def register(userType: UserType[?]): Unit = registry.put(userType.typeName, userType)
  
  def getTypeNameList: List[String] = registry.keys.toList
  
  def getBuilderByName(name: String): UserType[?] = {
    registry.get(name) match {
      case Some(userType) => userType
      case None => throw new IllegalArgumentException(s"Unknown type: $name")
    }
  }
}
